Selectbox theme additions Fixes #25965, #35246, #25700, #21193

cleidigh opened the pull request:
Selectbox theme additions Fixes #25965, #35246, #25700, #25619, #21193

I fixed all the issues I have found as well as addressing a lot of aesthetic issues.

Vertical margins addressed
Horizontal margins - controlled by ContextV - only flips when window edge hit
Could address this with optional margin requirements passed to ContextV (your call)
Cannot use CSS or forced resize because it conflicts with ContextV logic
Added VerticalScrollBar visibility option to listview - could also add horizontal
Tweaked margins for high contrast mode
DropDown is fully modal - could pass through other keyboard commands if desired
Re-Tested non-native select mode on : Windows 7, OS X, Debian
As first list/ContextV effort , make sure I didn't miss use anything !

See prior PR for more discussion #36193

bpasero commented:
@cleidigh good progress.

**Colors**
I am not so happy with the colors that are used yet:
* light theme: the active option shows in blue but then I would turn the foreground color of the text to white (same as we do in explorer). also, when I hover over the active option, the blue is removed and I am almost not seeing the slight hover feedback that shows up. also, we should probably use our default focus color (blue) for the border of the dropdown:
![image](https://user-images.githubusercontent.com/900690/32369899-61fafb68-c08b-11e7-8c7d-1573a06cc73e.png)
* dark theme: same issue with loosing the blue of the active option once I hover

**Accessibility:**
* when I press Space on the focused select widget, I would expect the element to expand and show all options.
* when I arrow up and down through the options, a screen reader is not reading out the elements

cleidigh commented:
@bpasero 


Addressed all comments, but a couple of questions/comments.  

Theme colors:
- Fixed registered colors to use white foreground color for light/dark themes, exceptions noted below.
- Tweaked hover background for select dd on light themes (normal list color has basically zero contrast)
- Did NOT change normal list colors - not sure if this is one of the things you referenced, exists currently.

Hover background behavior:
- hover color now supersedes select background colors
- I think this is what you mention your comments HOWEVER this is not the current behavior for list/quick open et cetera
- If I misinterpreted your comment, let me know and I will restore

Abyss
- Match list colors

Dark (VSC)
- Set foreground color to white
- Match focus border color

Dark+ (default dark)
- Set foreground color to white
- Match focus border color

High Contrast
- Scaled outline styles
- Slightly increased padding to make outlines more visible

Kimbie Dark
- Match list colors

Light (VSC)
- Set foreground color to white
- Match focus border color
- Increase hover contrast
  "dropdown.optionHoverBackground": "#e4e4e4"

Light (default light)

- Set foreground color to white
- Match focus border color
- Increase hover contrast
 "dropdown.optionHoverBackground": "#e4e4e4"

Monokai
- Match list colors

Monokai Dimmed
- Match list colors
- NOTE: The colors for active selection, focus, hover are all the same giving zero Contrast
- Should these be changed?

Quiet Light
- Match list colors
- Match focus border color
- Set foreground color to main foreground color (white not enough contrast)
    "dropdown.optiooverBackground": "#e4e4e4"

Red
- Match list colors

Solarized Light
- Match list colors

Solarized Dark
- Match list colors

Tomorrow Night Blue
- Match list colors

Accessibility/Behaviors:
- Added ARIA labels for select options
- Installed NVDA - You should see Dragon arguing with the screen reader  ;-) !
- Seems to read off labels okay

- Select dropdown expand with Space - just figured out you meant for native mode
- I restored original, native drop down with Space (non-native had already)
- Added Enter expansion as well - as in original

Saludos

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.css:
```diff
@@ -6,4 +6,63 @@
+}
+
+.monaco-shell.hc-black .select-dropdown-container .selectbox-dropdown-list-container {
+	padding-bottom: 4px !important;
```
@cleidigh not a big fan of using `!important` in CSS unless there is no other way. Is this really needed?

cleidigh replied:
Done.

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.css:
```diff
@@ -6,4 +6,63 @@
+	overflow: hidden !important;
+}
+
+.monaco-shell .select-dropdown-container .selectbox-dropdown-list-container {
```
@cleidigh I think a hierarchy should be expressed in CSS if clearly present. E.g. `.select-dropdown-container .selectbox-dropdown-list-container` can be `.select-dropdown-container > .selectbox-dropdown-list-container` imho

cleidigh replied:
Done.

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.ts:
```diff
@@ -5,56 +5,190 @@
+		// Hardcode native select switch based on platform
+		switch (process.platform) {
+			case 'darwin':
+				this._useNativeSelect = true;
```
@cleidigh since this file is growing rather large because it contains both implementations of native and custom rendering I suggest to extract the two into files of their own and reference them from this class? E.g. `selectBoxNative.ts` and `selectBoxCustom.ts` which then would be used from `selectBox.ts` depending on the OS. I think that makes maintainability easier in the future.

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.ts:
```diff
@@ -5,56 +5,190 @@
+	private selectList: List<ISelectOptionItem>;
+	private selectDropDownListContainer: HTMLElement;
+	private widthControlElement: HTMLElement;
+	// private widthControDivElement: HTMLElement;
```
@cleidigh the `widthControDivElement` is not used?

cleidigh replied:
Done

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.ts:
```diff
@@ -5,56 +5,190 @@
+		// Setup list for drop-down select
+		this.createSelectList(this.selectDropDownContainer);
+
+		// TODO:cleidigh - find better width control
```
@cleidigh can you fix or remove this TODO?

cleidigh replied:
Removed

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.ts:
```diff
@@ -5,56 +5,190 @@
+		this.widthControlElement.className = 'option-text-width-control';
+		dom.append(widthControlInnerDiv, this.widthControlElement);
+		// this.widthControDivElement = widthControlInnerDiv;
+		// TODO:cleidigh - find better width control
```
@cleidigh can you fix or remove this TODO?

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.ts:
```diff
@@ -115,31 +333,87 @@ export class SelectBox extends Widget {
+		if (!this._useNativeSelect) {
+
+			if (this.styles.selectOptionCheckedBackground) {
+				content.push(`.monaco-shell .select-dropdown-container .selectbox-dropdown-list-container .monaco-list .monaco-list-row.focused:not(:hover) { background-color: ${this.styles.selectOptionCheckedBackground} !important; }`);
```
@cleidigh again please express the DOM hierarchy if possible, e.g. `.select-dropdown-container .selectbox-dropdown-list-container` => `.select-dropdown-container > .selectbox-dropdown-list-container`

cleidigh replied:
Done.

bpasero commented on src/vs/platform/theme/common/colorRegistry.ts:
```diff
@@ -183,6 +183,11 @@ export const inputValidationErrorBorder = registerColor('inputValidation.errorBo
 export const selectBackground = registerColor('dropdown.background', { dark: '#3C3C3C', light: Color.white, hc: Color.black }, nls.localize('dropdownBackground', "Dropdown background."));
 export const selectForeground = registerColor('dropdown.foreground', { dark: '#F0F0F0', light: null, hc: Color.white }, nls.localize('dropdownForeground', "Dropdown foreground."));
 export const selectBorder = registerColor('dropdown.border', { dark: selectBackground, light: '#CECECE', hc: contrastBorder }, nls.localize('dropdownBorder', "Dropdown border."));
+export const selectOptionCheckedForeground = registerColor('dropdown.optionCheckedForeground', { dark: Color.white, light: Color.white, hc: null }, nls.localize('dropdownOptionCheckedForground', "Dropdown option checked foreground."));
```
@cleidigh actually thinking more about it, why do we have these colors at all? can we not just use the tree colors? since we are already using the same widget I see no much value in having different colors only for the dropdown....

cleidigh replied:
Using list colors.

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.ts:
```diff
@@ -152,4 +426,243 @@ export class SelectBox extends Widget {
+				if (this.options[index].length > this.options[longest].length) {
+					longest = index;
+				}
+			};
```
@cleidigh extra comma to remove

cleidigh replied:
Done

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.ts:
```diff
@@ -71,21 +206,104 @@ export class SelectBox extends Widget {
+
+		this.toDispose.push(dom.addDisposableListener(this.selectElement, dom.EventType.CLICK, () => {
+			this.showSelectDropDown();
+			event.preventDefault();
```
@cleidigh I suggest to use the shorter [`stop`](https://github.com/Microsoft/vscode/blob/master/src/vs/base/browser/dom.ts#L766) function with the option to cancel bubble.

bpasero commented on src/vs/base/browser/ui/selectBox/selectBox.ts:
```diff
@@ -152,4 +426,243 @@ export class SelectBox extends Widget {
+		return elementWidth;
+	}
+
+	public cloneElementFont(source: HTMLElement, target: HTMLElement) {
```
@cleidigh not so happy:
* why is the method `public`?
* it uses `var`, which we try to avoid (e.g. use `let` or `const` instead)
* can the value be cached and reused instead of always asking for it?

cleidigh commented:
@bpasero 
Thanks for the review comments.  I will work on these as fast as I can today.
Not sure about the cleanest way to split the class across files.  I will investigate/question if necessary.

cleidigh commented on src/vs/workbench/parts/terminal/electron-browser/terminalActions.ts:
```diff
@@ -430,9 +431,10 @@ export class SwitchTerminalInstanceActionItem extends SelectActionItem {
 		@ITerminalService private terminalService: ITerminalService,
-		@IThemeService themeService: IThemeService
+		@IThemeService themeService: IThemeService,
+		@IContextViewService private contextViewService: IContextViewService
```
Done

cleidigh commented on src/vs/workbench/parts/output/browser/outputActions.ts:
```diff
@@ -108,9 +109,10 @@ export class SwitchOutputActionItem extends SelectActionItem {
 		@IOutputService private outputService: IOutputService,
-		@IThemeService themeService: IThemeService
+		@IThemeService themeService: IThemeService,
+		@IContextViewService private contextViewService: IContextViewService
```
Done

cleidigh commented on src/vs/workbench/parts/debug/browser/debugActionItems.ts:
```diff
@@ -40,10 +41,11 @@ export class StartDebugActionItem implements IActionItem {
 		@IConfigurationService private configurationService: IConfigurationService,
-		@ICommandService private commandService: ICommandService
+		@ICommandService private commandService: ICommandService,
+		@IContextViewService private contextViewService: IContextViewService,
```
Done

Is this new warning?  I don't think I saw this before.

bpasero replied:
@cleidigh yes we added lots of more warning support recently by adopting TSLint 5

cleidigh commented on src/vs/workbench/parts/debug/browser/debugActionItems.ts:
```diff
@@ -184,9 +186,10 @@ export class FocusProcessActionItem extends SelectActionItem {
 		@IDebugService private debugService: IDebugService,
-		@IThemeService themeService: IThemeService
+		@IThemeService themeService: IThemeService,
+		@IContextViewService private contextViewService: IContextViewService
```
Done

cleidigh commented:
Selection color precedence over hover:

I had understood the opposite before, my bad...
Fixed.

bpasero commented:
@cleidigh awesome. take your time, I think with the stuff addressed we can merge this in for December milestone. 

Splitting the custom part out of the native one makes it easier for me to review the change. Now it is very hard to understand the impact for the native implementation. I think stuff will be cleaner when the two implementations are separated.

cleidigh commented:
@bpasero 
I basically have two things left to deal with, colors and file separation.
I originally ran into some problems trying to just use the list colors, I will see if I still have problems.

For the class separation: I am looking into possible design patterns and want your input before I cut things up.

Most basic:
- Keep Main widget methods (render, layout, style etc) in selectBox class/file 
- Have these call equivalent methods in each of the native and list files

Mixin Methods:
- Create classes/files with mixing from native and list classes/files
- Use applyMixins function to mixed methods from appropriate selectbox type per OS
- Must have placeholder methods in the main select box file
- Cleaner since does not require if/else for every method to call the sub methods

Preference or other option?

cleidigh commented:
@bpasero 
December sounds good, I knew it was getting too close for November.

I have one other problem - I am trying to see the problem with the debug drop-down. 
I cannot get code OSS to do Debug, says cannot to node or Chrome.
is this disabled code OSS?  if so any workaround?

cleidigh commented:
Cool I will try that.

Any preference on the separation approach?

bpasero commented:
@cleidigh well maybe the `selectBox.ts` could wrap around a `ISelectBox` interface that is implemented both for custom and for native. These two implementations can live next to each other in separate files. Within the constructor of `selectBox.ts` we do the check for the platform and then instantiate either the one or the other. But throughout the rest of the file we only talk to the interface.

cleidigh commented:
@bpasero 
I think we are close, but if I just do an interface without the mixing I don't think I can have 
method overrides for the widget functions which I believe are expected to be included
in the main select box class.  if I do mixing overrides I can get it to work although there is still a couple approaches to this as well.

bpasero commented:
@cleidigh I think it would work with an interface. basically you are delegating the calls to selectBox.ts into the implementation through the interface. So depending on which implementation is enabled it will reach either native or custom implementation.

cleidigh commented:
@bpasero 
I have most things fixed.
I cannot get Code OSS to do debugging to test the problem with the debugger widget.
I did not have the debug extension folders, I think they are integrated.
I had to clone and build the two debugger extensions.  I think that worked they seem to be loaded.
Cannot launch a debug session I get a connection error.
Also tried to package OSS thinking that might be more direct, packaging builds okay but does not run ,two processes are started but nothing appears on screen.  
I am at a loss to continue.

bpasero commented:
@cleidigh maybe easier if you try to debug an extension because everything is setup for you there. Just:
* `npm install -g yo`
* `npm install -g generator-code`
* `yo code`
* pick the JS extension
* do this two times to have multiple debug targets
* open both extensions in one window
* run both from the debug viewlet

cleidigh commented:
@bpasero 
Thanks that worked.  Looks like it works I got the drop down, maybe it was a merge issue.
I will commit and post when I have everything the separation done just to have some work in progress
before I re- factor everything.

cleidigh commented on src/vs/platform/theme/common/styler.ts:
```diff
@@ -101,13 +101,29 @@ export interface ISelectBoxStyleOverrides extends IStyleOverrides {
+	selectOptionHoverBackground?: ColorIdentifier;
+	selectOptionCheckedOutline?: ColorIdentifier;
+	selectOptionHoverOutline?: ColorIdentifier;
+	// listFocusForeground?: ColorIdentifier;
```
Fixed with list color change.

cleidigh commented on extensions/theme-abyss/themes/abyss-color-theme.json:
```diff
@@ -281,6 +281,9 @@
 		"progressBar.background": "#0063a5",
 
 		"dropdown.background": "#181f2f",
+		"dropdown.optionCheckedBackground": "#08286b",
```
Use list color definitions

cleidigh commented on src/vs/platform/theme/common/colorRegistry.ts:
```diff
@@ -183,6 +183,11 @@ export const inputValidationErrorBorder = registerColor('inputValidation.errorBo
 export const selectForeground = registerColor('dropdown.foreground', { dark: '#F0F0F0', light: null, hc: Color.white }, nls.localize('dropdownForeground', "Dropdown foreground."));
 export const selectBorder = registerColor('dropdown.border', { dark: selectBackground, light: '#CECECE', hc: contrastBorder }, nls.localize('dropdownBorder', "Dropdown border."));
+export const selectOptionCheckedForeground = registerColor('dropdown.optionCheckedForeground', { dark: Color.white, light: Color.white, hc: null }, nls.localize('dropdownOptionCheckedForground', "Dropdown option checked foreground."));
+export const selectOptionCheckedBackground = registerColor('dropdown.optionCheckedBackground', { dark: '#094771', light: '#3399FF', hc: null }, nls.localize('dropdownOptionCheckedBackground', "Dropdown option checked background."));
```
Use list color definitions.

cleidigh commented:
@bpasero 
I believe I have addressed all your issues.  To use your expression, I am not "happy" with the separation
approach, but after trying a variety of other ways including decorators and mixins , this seemed to be the best way.

Everything should be pretty clean now, I now use all list colors and passed focusBorder directly into
SelectBoxStyles.

Hopefully it looks good to you...

cleidigh commented:
Missed a couple things :-1: 
- var -> let
- duplicated listener

bpasero commented:
@cleidigh fyi did some last polish via https://github.com/Microsoft/vscode/pull/37533/commits/6f14fd3515bff947a3185084a83619c8a194f1fa
* properly hooked up the dispose method to the delegate
* separated CSS according to implementations
* removed some unnecessary null checks for the delegate

bpasero commented:
@cleidigh thanks for the work here 👍

cleidigh commented:
@bpasero 
Excellent !   Very glad this could get merged ;-)  Will address #40438 ASAP
I reviewed your polish to learn and understand.  For my edification -
- I see you want to do the disposal within the class not the client - I assume this is the normal approach.
- Did you remove the null checks because all code paths essentially guarantee SelectBoxIDelegate is set?

I appreciate you working with me on this for all this time.  Really learned a lot for future contributions !
Cheers

bpasero commented:
@cleidigh ah yeah I did not see that the disposal thing was passed over. I think I still prefer the style where the disposing is passed on to the child element, that is what we typically do in other places.

And yeah, the `SelectBoxIDelegate` can never be null so removing some code that would never get triggered.
